Update passkey form autofill article (#9196)
GoogleChrome/web.dev: src/site/content/en/identity/passkey-form-autofill/index.md

[Passkeys](https://developers.google.com/identity/passkeys) replace passwords
and make user accounts on the web safer, simpler, easier to use. However, the
transition from password-based to passkey-based authentication can complicate

> the user experience. Leveraging form autofill to suggest passkeys can help

Reviewer: Personally, I prefer use over leverage—is there any specific reason to change this phrasing?

Author: ```suggestion
the user experience. Using form autofill to suggest passkeys can help
```

create a unified experience.

## Why use form autofill to sign-in with a passkey?

> With a passkey, a user can sign in to a website just by using fingerprint, face

Reviewer: Just to check—I'm aware it was the same copy before as well—but is it meant to be a face pin or a device pin, because if not, an Oxford comma might come in handy here. I'm guessing it's implying face identification?

Author: It's about a fingerprint biometric sensor, face recognition or a device PIN. So PIN only applies to the device. I think Oxford comma sounds suitable in this case.

```suggestion
With a passkey, a user can sign in to a website just by using fingerprint, face,
```

can include passkeys in autofill suggestions. This is called [conditional
UI](https://github.com/w3c/webauthn/wiki/Explainer:-WebAuthn-Conditional-UI) and
it's a part of [the WebAuthn standard](https://w3c.github.io/webauthn/).

<figure>
  {%
    Video src="video/YLflGBAPWecgtKJLqCJHSzHqe2J2/ZaxgMtkOh5CX2B1ZT9gd.mp4",
    autoplay="true", loop="true"
  %}
</figure>

As soon as the user taps on the username input field, an autofill suggestion
dialog pops up which highlights the stored passkeys along with password autofill
suggestions. The user can then choose an account and use the device screen lock
to sign in.

This way, users can sign in to your website with the existing form as if nothing
has changed, but with [the added security benefit of
passkeys](https://developers.google.com/identity/passkeys#security-considerations)
if they have one.

{% Aside 'success' %}

When a user signs in with a password, consider using that opportunity to offer
them to [create a passkey](/passkey-registration).

{% endAside %}

## How it works

To authenticate with a passkey, you use [the WebAuthn 
API](https://w3c.github.io/webauthn/).

> In a passkey authentication flow, four participants are involved in addition to 

Reviewer: Would it make more sense to describe this as "The four components in a passkey authentication flow are:"?

Do we have to mention the user here?

Author: Four components sounds good to me. I don't have strong opinion on if adding a user is a good idea or not, but I omitted it because it was obvious.

```suggestion
The four components in a passkey authentication flow are:
```

Feel free to modify the suggesion and resolve.

the user:

> * **Backend**: Your accounts database or backend server that stores the public 

Reviewer: accounts database? Like a database holding all account information? I guess I'm confused by the possessive "your" here—is the developer reading this in charge of the database or the server?

Author: Yeah, this should be "your backend server that holds accounts database"

Author: Added a suggestion.

> * **Backend**: Your accounts database or backend server that stores the public 
>   key and other metadata about the passkey.

Author: ```suggestion
* **Backend**: Your backend server that holds accounts database where it stores the public 
  key and other metadata about the passkey.
```

> * **Frontend**: Your frontend server which communicates with it and serves 
>   requests.

Author: ```suggestion
* **Frontend**: Your frontend which communicates with the browser and sends fetch 
  requests to the backend.
```

> * **Frontend**: Your frontend server which communicates with it and serves 

Author: Added a suggestion to clarify this.

* **Browser**: The user's browser which is running your Javascript.
* **Authenticator**: The user's authenticator which creates and stores the 
  passkey. This may be on the same device as the browser (e.g. when using 
  Windows Hello) or on another device, like a phone.

<figure class="screenshot">
  {%
    Img src="image/YLflGBAPWecgtKJLqCJHSzHqe2J2/vAwnMiLi7U1spDopYg2Q.png",
    alt="Passkey authentication diagram", width="800", height="254"
  %}
</figure>

> 1. As soon as a user lands on the frontend, it requests a challenge from the 

Reviewer: I think I saw a front-end elsewhere—how are we writing it, frontend or front-end?

Is it the server requesting a challenge? Can we specify that here?

Author: Latter is correct. Frontend requests to fetch a challenge from the backend.
I'd stick with "frontend".

Author: A change made to the "frontend" explanation should have clarified this question.

   backend to authenticate with a passkey and calls 
   `navigator.credentials.get()` to initiate authenticating with a passkey. 

>    This returns a promise.

Reviewer: Do we define what a "promise" is anywhere?

Author: ```suggestion
   This returns a `Promise`.
```

* The browser supports WebAuthn.
* The browser supports WebAuthn conditional UI.

```js  
// Availability of `window.PublicKeyCredential` means WebAuthn is usable.  
if (window.PublicKeyCredential &&  
    PublicKeyCredential.​​isConditionalMediationAvailable) {  
  // Check if conditional mediation is available.  
  const isCMA = await PublicKeyCredential.​​isConditionalMediationAvailable();  
  if (isCMA) {  
    // Call WebAuthn authentication  
  }  
}  
```

### Fetch a challenge from the RP server

Fetch a challenge from the RP server that is required to call 
`navigator.credentials.get()`:

* **[`challenge`](https://w3c.github.io/webauthn/#dom-publickeycredentialcreationoptions-challenge)**: 
  A server-generated challenge in an ArrayBuffer. This is required to prevent 
  replay attacks. Make sure to generate a new challenge on every sign-in attempt 
  and disregard it after a certain duration or after a sign-in attempt fails to 
  validate. Consider it like a CSRF token.

{% Aside %}

ArrayBuffer values transferred from the server such as `challenge`, `user.id`
and credential `id` for `excludeCredentials` need to be encoded on transmission.

> Make sure to decode them on the RP front-end before passing the values to the WebAuthn

Author: ```suggestion
Make sure to decode them on the RP frontend before passing the values to the WebAuthn
```
